src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs

```diff
@@ -102,6 +106,67 @@ public override async Task<bool> CanChangeNamespaceAsync(Document document, Synt
             return !applicableContainers.IsDefault;
         }
 
+        public override string TryBuildNamespaceFromFolders(IEnumerable<string> folders, ISyntaxFacts syntaxFacts)
+        {
+            var parts = folders.SelectMany(folder => folder.Split(new[] { '.' })).SelectAsArray(EscapeIdentifier);
+            return parts.All(syntaxFacts.IsValidIdentifier) ? string.Join(".", parts) : null;
+        }
+
+        public override async Task<Solution> ChangeTopLevelNamespacesAsync(
+            Document document,
+            string targetNamespace,
+            CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Don't descend into anything other than top level declarations from the root.
+            // ChangeNamespaceService only controls top level declarations right now.
+            // Don't use namespaces that already match the target namespace
+            var namespaceDeclaration = syntaxRoot
+                .DescendantNodes(n => !syntaxFacts.IsDeclaration(n))
+                .FirstOrDefault(n => syntaxFacts.IsNamespaceDeclaration(n));
+
+            if (namespaceDeclaration is null)
+            {
+                throw new InvalidOperationException("Unable to find valid top level namespaces to change");
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var originalNamespaceName = semanticModel.GetDeclaredSymbol(namespaceDeclaration).Name;
+
+            document = document.WithSyntaxRoot(syntaxRoot);
+
+            var solution = document.Project.Solution;
+
+            while (true)
+            {
+                syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var namespaceToRename = await GetMatchingTopLevelNamespaceAsync(document, originalNamespaceName, cancellationToken).ConfigureAwait(false);
+
+                if (namespaceToRename is null)
+                {
+                    break;
+                }
+
+                solution = await ChangeNamespaceAsync(document, namespaceToRename, targetNamespace, cancellationToken).ConfigureAwait(false);
+                document = solution.GetRequiredDocument(document.Id);
+            }
```

Comment:
why do we need a while loop here?

Reply:
It continues to find all top level namespaces named the same in the document. I tried to do with annotations originally, but the `ChangeNamespaceAsync` was getting rid of them and it was becoming a very complex change to not have that happen.

Reply:
1. got it.  
2. the infinite loop scares me.  are we certain that calls like ChangeNamespaceAsync always succeed and will change teh name?

Basically, i think we need a way to cap this.

Reply:
Good point. We can cap by the number of top level namespaces in the original document since that won't change